sql-plugin/src/main/scala/com/nvidia/spark/rapids/literals.scala

```diff
@@ -146,16 +124,218 @@ object GpuScalar {
+ * it is used only on the device side (This is the ideal case we like, since all is on the GPU).
+ * And only need to copy the data to the host once if it is used on the host side.
+ *
+ * So a GpuScalar will have at most one data copy but support all the cases. No round-trip
+ * happens.
+ *
+ * Another reason why storing the Scala value in addition to the cudf Scalar is
+ * `GpuDateAddInterval` and 'GpuTimeMath' have different algorithms with the 3 members of
+ * a `CalendarInterval`, which can not be supported by a single cudf Scalar now.
+ *
+ * Do not create a GpuScalar from the constructor, instead call the factory APIs above.
+ */
+class GpuScalar private(
+    private var scalar: Option[Scalar],
+    private var value: Option[Any],
+    val dataType: DataType) extends Arm with AutoCloseable {
+
+  if(scalar.isEmpty && value.isEmpty) {
+    throw new IllegalArgumentException("GpuScalar requires at least a value or a Scalar")
+  }
+  if (value.isDefined && value.get.isInstanceOf[Scalar]) {
+    throw new IllegalArgumentException("Value should not be Scalar")
+  }
+
+  /**
+   * Gets the internal cudf Scalar of this GpuScalar.
+   *
+   * This will not increase any reference count. So users need to close either the GpuScalar or
+   * the return cudf Scalar, not both.
+   */
+  def getBase: Scalar = {
+    if (scalar.isEmpty) {
+      scalar = Some(GpuScalar.from(value.get, dataType))
+    }
+    scalar.get
+  }
+
+  /**
+   * Gets the internal Scala value of this GpuScalar.
+   */
+  def getValue: Any = {
+    if (value.isEmpty) {
+      value = Some(GpuScalar.extract(scalar.get))
     }
+    value.get
   }
+
+  /**
+   * GpuScalar is valid when
+   *   the Scala value is not null if it is defined, or
+   *   the cudf Scalar is valid if the Scala value is not defined.
+   * Because a cudf Scalar created from a null is invalid.
+   */
+  def isValid: Boolean = value.map(_ != null).getOrElse(getBase.isValid)
+
+  /**
+   * Whether the GpuScalar is Nan. It works only for float and double types, otherwise
+   * an exception will be raised.
+   */
+  def isNan: Boolean = dataType match {
+    case FloatType => getValue.asInstanceOf[Float].isNaN
+    case DoubleType => getValue.asInstanceOf[Double].isNaN
+    case o => throw new IllegalStateException(s"$o is doesn't support NaNs")
+  }
+
+  /**
+   * Whether the GpuScalar is not a Nan. It works only for float and double types, otherwise
+   * an exception will be raised.
+   */
+  def isNotNan: Boolean = !isNan
+
+  /**
+   * Increment the reference count for this scalar. You need to call close on this
+   * to decrement the reference count again.
+   */
+  def incRefCount: this.type = incRefCountInternal(false)
+
+  override def close(): Unit = {
+    this.synchronized {
+      refCount -= 1
+      if (refCount == 0) {
+        scalar.foreach(_.close())
+        scalar = Some(null)
+        value = Some(null)
```

Comment:
I would expect these to be `None` or just `null`.  (`null` would be to help trigger an NPE if someone tries to do something with the instance after it's closed.). Curious about the rationale for `Some(null)`.

Reply:
I just wanted to let the value/scalar be null. But your suggestion of Option being null' is much better.
Updated.